lib/ddtrace/contrib/rack/rum_injection.rb

```diff
@@ -0,0 +1,239 @@
+            (headers[CONTENT_TYPE_HEADER].include?(HTML_CONTENT) ||
+            headers[CONTENT_TYPE_HEADER].include?(XHTML_CONTENT))
+        end
+
+        def attachment?(headers)
+          (headers && headers.key?(CONTENT_DISPOSITION_HEADER) && !headers[CONTENT_DISPOSITION_HEADER].nil?) &&
+            !headers[CONTENT_DISPOSITION_HEADER].include?(INLINE)
+        end
+
+        def streaming?(headers, env)
+          # https://api.rubyonrails.org/classes/ActionController/Streaming.html
+          # rails recommends disabling middlewares that interact with response body
+          # when streaming via ActionController::Streaming
+          # TODO: if required to patch streaming, investigate patching further upstream, in the render action perhaps
+          return true if (headers && headers.key?(TRANSFER_ENCODING_HEADER) &&
+                         headers[TRANSFER_ENCODING_HEADER] == TRANSFER_ENCODING_CHUNKED) ||
+                         (
+                            headers && headers.key?(CONTENT_TYPE_HEADER) &&
+                            !headers[CONTENT_TYPE_HEADER].nil? &&
+                            headers[CONTENT_TYPE_HEADER].start_with?(CONTENT_TYPE_STREAMING)
+                         )
+
+          # if we detect Server Side Event streaming controller, assume streaming
+          defined?(ActionController::Live) &&
+            env[ACTION_CONTROLLER_INSTANCE].class.included_modules.include?(ActionController::Live)
+        end
+
+        def no_cache?(headers, env)
+          # TODO: this is very complex, is there an easier way to determine cache behavior on cdn and browser
+          # TODO: glob performance may be worse than regex
+          return false unless configuration[:rum_cached_pages].none? do |page_glob|
+            File.fnmatch(page_glob, env['PATH_INFO']) if env['PATH_INFO']
+          end
+
+          return false unless headers
+
+          # first check Surrogate-Controle, which fastly uses
+          # indicates a cdn cache if Surrogate-Control max-age>0,
+          # otherwise check cache-control, then expiry to determine if there is browser cache
+          return false if surrogate_cache?(headers)
+
+          # then check Cache-Control
+          if headers.key?(CACHE_CONTROL_HEADER) && !headers[CACHE_CONTROL_HEADER].nil?
+            # s-maxage gets priority over max-age since s-maxage sits at cdn level
+            # cdn cache if s-maxage >0,
+            # otherwise check max-age, (no-store|no-cache|private) or expires to determine if therre is browser cache
+            return false if server_cache?(headers)
+
+            # then check max-age
+            if headers[CACHE_CONTROL_HEADER].include?(MAX_AGE)
+              # only not cached if max-age is 0
+              return headers[CACHE_CONTROL_HEADER].include?(MAX_AGE_ZERO)
+            end
+
+            # not cached if marked no-store, no-cache, or private
+            return (headers[CACHE_CONTROL_HEADER].include?(NO_CACHE) ||
```

Comment:
We should check these hard conditions first, as they are easier to help us short circuit out of this method.

Reply:
see other comment re: the headers have an intentional priority ordering. I've update the code comments to describe this more appropriately for future developers